### PR TITLE
pkg/kvstore: fixed race in internal mutex map

### PR DESCRIPTION
**Pull request by aanm**
Signed-off-by: André Martins <andre@cilium.io>

### CODE REVIEW

File: pkg/kvstore/lock.go [hunk 1]
```diff
@@ -38,22 +38,28 @@ type Lock struct {
...
 	lockPathsMU.Unlock()
 
 	trace("Creating lock", nil, logrus.Fields{fieldKey: path})
 
 	// Take the local lock as both etcd and consul protect per client
-	lockPaths[path].Lock()
+	localLock.Lock()
 
 	lock, err := Client().LockPath(path)
 	if err != nil {
```
Comment (raybejjani) [inline pkg/kvstore/lock.go]: we need to use localLock in the line below this (which I can't comment on in github :/ ) when we release it in the error case.

File: pkg/kvstore/lock.go [hunk 2]
```diff
@@ -38,22 +38,28 @@ type Lock struct {
...
+	lockPathsMU lock.RWMutex
 	lockPaths   = map[string]*lock.Mutex{}
 )
 
 // LockPath locks the specified path and returns the Lock
 func LockPath(path string) (*Lock, error) {
+	var (
+		localLock *lock.Mutex
+		ok        bool
+	)
```
Comment (raybejjani) [inline pkg/kvstore/lock.go]: Not a big deal but why declare these here? you could `:=` below since we don't put things into the if line.

**Review by raybejjani — CHANGES_REQUESTED** [review]

File: pkg/kvstore/lock.go (hunk 2 above)
Comment (aanm) [reply pkg/kvstore/lock.go]: Correct, thanks! Fixed

File: pkg/kvstore/lock.go (hunk 1 above)
Comment (aanm) [reply pkg/kvstore/lock.go]: Correct, thanks! Fixed

**Review by raybejjani — APPROVED** [review]

File: pkg/kvstore/lock.go [hunk 3]
```diff
@@ -67,9 +69,16 @@ func LockPath(path string) (*Lock, error) {
 
 // Unlock unlocks a lock
 func (l *Lock) Unlock() error {
+	// Unlock kvstore mutex first
 	err := l.lock.Unlock()
 
-	lockPaths[l.path].Unlock()
+	lockPathsMU.RLock()
+	localLock, ok := lockPaths[l.path]
```
Comment (tgraf) [inline pkg/kvstore/lock.go]: Why is this change needed?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/kvstore/lock.go (hunk 3 above)
Comment (aanm) [reply pkg/kvstore/lock.go]: We are reading from `lockPaths` so we need to have a read lock to protect that "read"

Comment (tgraf) [reply pkg/kvstore/lock.go]: Why do we need to re-lookup the lock variable?

Comment (tgraf) [reply pkg/kvstore/lock.go]: Discussed offline, the changes allow to cleanup the map

Comment (tgraf) [reply pkg/kvstore/lock.go]: Discussed offline, the changes allow to cleanup the map

**Review by tgraf — APPROVED** [review]

**Review by ianvernon — APPROVED** [review]

**Comment by ianvernon** [comment]
What exactly was the race condition that was observed?

**Comment by aanm** [comment]
>What exactly was the race condition that was observed?

We were reading from the `lockPaths` without doing an Lock
